metrics/core/src/data.rs

```diff
@@ -0,0 +1,875 @@
+            network_mempool_size: b.network_mempool_size as f64,
+            network_transactions_per_second,
+            network_tip_hashes_count: b.network_tip_hashes_count as f64,
+            network_difficulty: b.network_difficulty,
+            network_past_median_time: b.network_past_median_time as f64,
+            network_virtual_parent_hashes_count: b.network_virtual_parent_hashes_count as f64,
+            network_virtual_daa_score: b.network_virtual_daa_score as f64,
+
+            data: b.clone(),
+        }
+    }
+}
+
+/// Display KB or KiB if `short` is false, otherwise if `short` is true
+/// and the value is greater than 1MB or 1MiB, display units using [`as_data_size()`].
+pub fn as_kb(bytes: f64, si: bool, short: bool) -> String {
+    let unit = if si { 1000_f64 } else { 1024_f64 };
+    if short && bytes > unit.powi(2) {
+        as_data_size(bytes, si)
+    } else {
+        let suffix = if si { " KB" } else { " KiB" };
+        let kb = bytes / unit; //(( * 100.) as u64) as f64 / 100.;
+        format_with_precision(kb) + suffix
+    }
+}
+
+/// Display MB or MiB if `short` is false, otherwise if `short` is true
+/// and the value is greater than 1GB or 1GiB, display units using [`as_data_size()`].
+pub fn as_mb(bytes: f64, si: bool, short: bool) -> String {
+    let unit = if si { 1000_f64 } else { 1024_f64 };
+    if short && bytes > unit.powi(3) {
+        as_data_size(bytes, si)
+    } else {
+        let suffix = if si { " MB" } else { " MiB" };
+        let mb = bytes / unit.powi(2); //(( * 100.) as u64) as f64 / 100.;
+        format_with_precision(mb) + suffix
+    }
+}
+
+/// Display GB or GiB if `short` is false, otherwise if `short` is true
+/// and the value is greater than 1TB or 1TiB, display units using [`as_data_size()`].
+pub fn as_gb(bytes: f64, si: bool, short: bool) -> String {
+    let unit = if si { 1000_f64 } else { 1024_f64 };
+    if short && bytes > unit.powi(4) {
+        as_data_size(bytes, si)
+    } else {
+        let suffix = if si { " GB" } else { " GiB" };
+        let gb = bytes / unit.powi(3); //(( * 100.) as u64) as f64 / 100.;
+        format_with_precision(gb) + suffix
+    }
+}
+
+/// Display units dynamically formatted based on the size of the value.
+pub fn as_data_size(bytes: f64, si: bool) -> String {
+    let unit = if si { 1000_f64 } else { 1024_f64 };
+    let mut size = bytes;
+    let mut unit_str = " B";
+
+    if size >= unit.powi(4) {
+        size /= unit.powi(4);
+        unit_str = " TB";
+    } else if size >= unit.powi(3) {
+        size /= unit.powi(3);
+        unit_str = " GB";
+    } else if size >= unit.powi(2) {
+        size /= unit.powi(2);
+        unit_str = " MB";
+    } else if size >= unit {
+        size /= unit;
+        unit_str = " KB";
+    }
+
+    format_with_precision(size) + unit_str
+}
+
+/// Format supplied value as a float with 2 decimal places.
+fn format_as_float(f: f64, short: bool) -> String {
+    if short {
+        if f < 1000.0 {
```

Comment:
Optional: I based my function on this, and you could consider applying it here (or something adjacent to this):

```
fn to_human_readable(mut f: f64, precision: usize, suffix: &str) -> String {
    let units = vec!["", "K", "M", "G", "T", "P", "E"];
    let mut found_unit = "";

    for unit in units {
        if f < 1000.0 {
            found_unit = unit;
            break;
        } else {
            f /= 1000.0
        }
    }
    
    format!("{f:.precision$} {}{}", found_unit, suffix)
}
```

Reply:
In metrics, this is a performance and format-sensitive fn.  The `to_human_readable()` that you suggested has a `vec![]` allocation (it should be an array `[]`), it also executes a loop that results in multiple branching conditions and multiple divisions.  The fn you pointed out in metrics (`format_as_float()`) does not have any allocations and does not have a loop as it uses branching with a single division...  So it has higher performance.  Also, the behavior of these is quite different as there is a `short` variable, which when not specified, will result in a fully formatted number and "thousands separator" commas.

In an external project that I am about to release, these functions are performance-sensitive (as they can be called a few hundred times per sec) and format-specific, so I am just making some effort to optimize them.

Sidenote: metrics formatting used in the metrics-core crate doesn't use this specific function much, instead it uses sister fns such as `as_kb()`, `as_md()` etc., and these are strategically placed onto various fields/properties where the actual metrics should be reflected in specific units.  There are cases (such as total bandwidth consumption) where say the value might be 1 Gb, but you want to see the actual number of bytes or kilobytes changing... as otherwise you just stop seeing the actual value change.  This is not really related to perf-metrics as this is a more bandwidth-consumption-related issue (albeit RAM usage does fall under this category).  Anyway, just wanted to mention this to be aware of.